### ISSUE TITLE
Request for Mathematical Formulas of Constraint Violation Evaluation Component

### ISSUE THREAD

**Issue by shanefeng123** [AI4OPT/ml4acopf_benchmark#4]
Hi,

I am currently working on integrating your described neural network verification specification into my code. While the first component of the network is straightforward to implement, I am finding it challenging to incorporate the second component due to its complexity in the ONNX format.

Specifically, I am referring to the evaluation of constraint violations that involves nonlinear functions such as sigmoid, quadratic, and trigonometric functions (sine and cosine). To ensure accurate integration and implementation in my code, could you kindly provide the explicit mathematical formulas or detailed expressions for this component?

Having these formulas would greatly facilitate the integration process and ensure correctness. I appreciate your help and look forward to your response.

Best,
Shane

**Comment by klamike** [comment]
Hi, thanks for your interest in the benchmark. The model was implemented using an early internal version of what is now [ML4OPF](https://github.com/AI4OPT/ML4OPF). You can reference the implementation of the constraints to get the formulae, or use the library directly if you're using PyTorch. Specific links for each quantity are also included below for completeness.

The overall forward method of the benchmark model looks like:

```python
def forward(self, pd, qd):
  # regular feed-forward NN
  pg, qg, vm, va = self.neural_network(torch.cat([pd, qd], dim=1))
  
  # constraint residuals
  pf, pt, qf, qt = self.violation.flows_from_voltage_bus(vm, va)
  thrm_1, thrm_2 = self.violation.thermal_residual(pf, pt, qf, qt)
  pbal, qbal = self.violation.balance_residual(pd, qd, pg, qg, vm, pf, pt, qf, qt, embed="dense_matrix")

  # cat them all together
  return torch.cat([pg, qg, vm, va, pf, pt, qf, qt, thrm_1, thrm_2, pbal, qbal], dim=1)
```

`pg, qg, vm, va` comes from the neural network. This should be straightforward from the ONNX file. IIRC we used a sigmoid bounding function, i.e. `sigmoid(x) * (U - L) + L` to enforce `L ≤ x ≤ U` for `pg, qg, vm`.

`pf, pt, qf, qt` comes from [`ACViolation.flows_from_voltage_bus`](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/formulations/ac/violation.py#L67) (math happens [here](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/functional/ac.py#L114-L169) and [here](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/functional/__init__.py#L62-L66)/[here](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/formulations/ac/violation.py#L14-L25))

`thrm_1, thrm_2` comes from [`ACViolation.thermal_residual`](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/formulations/ac/violation.py#L150) (math happens [here](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/functional/ac.py#L11-L53))

`pbal, qbal` comes from [`ACViolation.balance_residual`](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/formulations/ac/violation.py#L176) (math happens [here](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/functional/ac.py#L57-L96)). Note this uses the `"dense_matrix"` embed method, which is implemented [here](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/formulations/violation.py#L324-L410) and [here](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/functional/incidence.py#L163-L165). IIRC this was required to support ONNX.

The dataset was generated by (again, an early internal version of what is now) the Julia package [OPFGenerator](https://github.com/AI4OPT/OPFGenerator). The [ACOPF formulation in Julia](https://github.com/AI4OPT/OPFGenerator/blob/main/src/opf/acp.jl) is another place you will find the constraint expressions, specifically [these lines](https://github.com/AI4OPT/OPFGenerator/blob/0b60e649073fb349126f5873ecea8d29cc9f81ab/src/opf/acp.jl#L81-L135).

**Comment by shanefeng123** [comment]
Hi,

Thanks I'll have a look.

Best,
Shane
